aws/src/main/java/org/apache/iceberg/aws/dynamodb/DynamoDbCatalog.java

```diff
@@ -0,0 +1,613 @@
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Tasks;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+import software.amazon.awssdk.services.dynamodb.model.AttributeDefinition;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.services.dynamodb.model.BillingMode;
+import software.amazon.awssdk.services.dynamodb.model.ConditionalCheckFailedException;
+import software.amazon.awssdk.services.dynamodb.model.CreateTableRequest;
+import software.amazon.awssdk.services.dynamodb.model.Delete;
+import software.amazon.awssdk.services.dynamodb.model.DeleteItemRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableRequest;
+import software.amazon.awssdk.services.dynamodb.model.DescribeTableResponse;
+import software.amazon.awssdk.services.dynamodb.model.GetItemRequest;
+import software.amazon.awssdk.services.dynamodb.model.GetItemResponse;
+import software.amazon.awssdk.services.dynamodb.model.GlobalSecondaryIndex;
+import software.amazon.awssdk.services.dynamodb.model.KeySchemaElement;
+import software.amazon.awssdk.services.dynamodb.model.KeyType;
+import software.amazon.awssdk.services.dynamodb.model.Projection;
+import software.amazon.awssdk.services.dynamodb.model.ProjectionType;
+import software.amazon.awssdk.services.dynamodb.model.Put;
+import software.amazon.awssdk.services.dynamodb.model.PutItemRequest;
+import software.amazon.awssdk.services.dynamodb.model.QueryRequest;
+import software.amazon.awssdk.services.dynamodb.model.QueryResponse;
+import software.amazon.awssdk.services.dynamodb.model.ResourceNotFoundException;
+import software.amazon.awssdk.services.dynamodb.model.ScalarAttributeType;
+import software.amazon.awssdk.services.dynamodb.model.TableStatus;
+import software.amazon.awssdk.services.dynamodb.model.TransactWriteItem;
+import software.amazon.awssdk.services.dynamodb.model.TransactWriteItemsRequest;
+import software.amazon.awssdk.services.dynamodb.model.UpdateItemRequest;
+
+/**
+ * DynamoDB implementation of Iceberg catalog
+ */
+public class DynamoDbCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces, Configurable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DynamoDbCatalog.class);
+  private static final int CATALOG_TABLE_CREATION_WAIT_ATTEMPTS_MAX = 5;
+  static final Joiner COMMA = Joiner.on(',');
+
+  private static final String GSI_NAMESPACE_IDENTIFIER = "namespace-identifier";
+  private static final String COL_IDENTIFIER = "identifier";
+  private static final String COL_IDENTIFIER_NAMESPACE = "NAMESPACE";
+  private static final String COL_NAMESPACE = "namespace";
+  private static final String PROPERTY_COL_PREFIX = "p.";
```

Comment:
@jackye1995 is there a reason we wouldn't just use a map in DynamoDB for these properties (as opposed to creating a prefixed column)?

Seems like might want to split out some of the well known values to top level properties (like location), but general properties just stored as a map.  Thoughts?

Reply:
The main concern here is that table properties can get rather messy in that anyone can add values and there's really no limitation of name.  This would in turn make the catalog table schema rather messy since it would be the union of all of the table property key names.  

I feel like since we don't have any use cases specifically called out for secondary indexes, we should use a map to reflect the table properties.  If we identify cases where we want additional secondary indexes, we could come up with a way to "promote" certain table properties, but I don't want to speculate too much until we understand what those cases really are.

Reply:
> The main concern here is that table properties can get rather messy in that anyone can add values and there's really no limitation of name. This would in turn make the catalog table schema rather messy since it would be the union of all of the table property key names.

DynamoDB is a noSQL database. Similar to HBase and Cassandra, the table schema only defines the primary key (in DynamoDB case is the hash key and range key, which are the `identifier` and `namespace` columns), and all the other columns are not a part of the schema. There is no such a concept that the table schema has to be a union of all column values.

> I feel like since we don't have any use cases specifically called out for secondary indexes, we should use a map to reflect the table properties.

I have listed one use case that we encounter a customer request about regarding ownership. The goal is to avoid the need to keep iterating the implementation with those feature requests.

Reply:
It's been a while since I've used DynamoDB, but it used to be that the console viewer would show all columns (I recognize that dynamodb itself only requires hash and sort keys).  Maybe that's changed and it's not an issue.

For the use case that you mentioned though, there's nothing specific that iceberg supports with the secondary index so it's somewhat external to the catalog implementation.  I'm not fundamentally opposed, just trying to make sure we have good justification for taking this approach, which is somewhat less straightforward than keeping the properties as a map.

Reply:
> It's been a while since I've used DynamoDB, but it used to be that the console viewer would show all columns

I see, yes in the console you still see all the available columns (it is based on the rows it scans in the first batch). The number of columns shown will be propotional to the number of properties that the user choose to add to the tables.

> there's nothing specific that iceberg supports with the secondary index so it's somewhat external to the catalog implementation. I'm not fundamentally opposed, just trying to make sure we have good justification for taking this approach

Yes totally agree, I think this is actually the most important reason we chose this approach, because it provides much more flexibility for users to extend based on this single solution, without the need for us to add any additional DynamoDB specific logic such as creating new GSIs in the future in open source. That will be much harder to do because it needs to add auto-migration logic to existing users which would potentially impact production performance. The functional benefit overweights the cost that the UI shows a few more columns in my perspective.

Reply:
ok, sounds good.